crates/fuel-core/src/coins_query.rs

```diff
@@ -221,6 +274,168 @@ pub async fn random_improve(
     Ok(coins_per_asset)
 }
 
+pub async fn select_coins_to_spend(
+    CoinsToSpendIndexIter {
+        big_coins_iter,
+        dust_coins_iter,
+    }: CoinsToSpendIndexIter<'_>,
+    total: u64,
+    max: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+    batch_size: usize,
+) -> Result<Vec<CoinsToSpendIndexEntry>, CoinsQueryError> {
+    const TOTAL_AMOUNT_ADJUSTMENT_FACTOR: u64 = 2;
+    if total == 0 && max == 0 {
+        return Ok(vec![]);
+    }
+
+    let adjusted_total = total.saturating_mul(TOTAL_AMOUNT_ADJUSTMENT_FACTOR);
+
+    let big_coins_stream = futures::stream::iter(big_coins_iter).yield_each(batch_size);
+    let dust_coins_stream = futures::stream::iter(dust_coins_iter).yield_each(batch_size);
+
+    let (selected_big_coins_total, selected_big_coins) =
+        big_coins(big_coins_stream, adjusted_total, max, excluded_ids).await?;
+
+    if selected_big_coins_total < total {
+        return Ok(vec![]);
+    }
```

Comment:
It would be nice to return an error here=)

Reply:
The empty vector will be converted to `CoinsQueryError::InsufficientCoinsForTheMax` on the callsite, as you noticed here: https://github.com/FuelLabs/fuel-core/pull/2463/files#r1881319248

It's not the best solution, but returning this error from within the `fn select_coins_to_spend()` function would require the function to know the `asset_id` and currently it's asset agnostic (I'd like to keep it that way).

I was considering adding a new error variant (for example `CoinSelectionError`) and convert it to proper `CoinsQueryError` on the callsite, but this complicates stuff.

I think I'll probably replace `vec![]` with `None` and see how it fits the design.

Reply:
I use the approach when I return `None` and construct the proper error with all necessary data upstream - commit: 2c4006939925e227771ee2a2bd800bbf3bc1be5c

Reply:
You can get `AssetId` from the `last_selected_big_coin` which you have below=)

Reply:
That won't work in case we don't select any coins in `big_coins()`. That's also why I cannot get it from one of the `selected_big_coins`, at least not consistently.

Reply:
I decided to just pass asset id to the function, so it can now return proper error (with correct `asset_id` and `collected_amount`). Commit: 6740e7d2707ab976b5595ed9487c9a5fa178b148